### PR TITLE
build(publish): pubish cadl emitter along with autorest.csharp

### LINKED ISSUE #2902: build(cadl): change how we publish cadl emitter and generate PR on `azure-sdk-for-net`
**Describe the issue or request**

Right now, any change regarding cadl generation is not effectively applied to `azure-sdk-for-net`, because:
- we don't automatically publish a new version of cadl emitter
  - we have to update the version of cadl emitter and run a separate pipeline to publish it
- we don't automatically bump the emitter version in `azure-sdk-for-net`

So here is the current process to apply the new version of emitter to `azure-sdk-for-net`:
- update the version of cadl emitter in [`src/CADL.Extension/Emitter.Csharp/package.json`](https://github.com/Azure/autorest.csharp/blob/1ca6314ddf96ae913b1993c7284b6991f10e26a9/src/CADL.Extension/Emitter.Csharp/package.json#L3)
- run [cadl.csharp - ci](https://dev.azure.com/azure-sdk/internal/_build?definitionId=5199&_a=summary) and approve it
- bump the version in `azure-sdk-for-net`, e.g. the [csharp-emitter-package.json](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/csharp-emitter-package.json) and re-generate the codes=

**Describe your ideas for solutions**

I think ultimately, we can do the same thing as `autorest.csharp` nuget package in [`AutoRest C# - Publish` pipeline](https://dev.azure.com/azure-sdk/internal/_build?definitionId=1375).
- automatically publish a new version of emitter npm package for each change
  - this is like [`@autorest/csharp-v3` package](https://www.npmjs.com/package/@autorest/csharp-v3), its version is like `3.0.0-beta.20201217.1`, but if you look into the source codes, its version is [`3.0.0`](https://github.com/Azure/autorest.csharp/blob/1ca6314ddf96ae913b1993c7284b6991f10e26a9/src/AutoRest.CSharp/package.json#L2)
  - one optimization we could do is that only publish new versions if there is any change under `src` folder
- automatically bump the emitter version in `azure-sdk-for-net`

### PR DESCRIPTION
**Pull request by archerzz**
- add cadl emitter build/test steps in exsiting autorest.csharp publish pipeline
- modify autorest.csharp publish script to publish the NPM package of cadl emitter `@azure-tool/cadl-csharp`
- modify the script to update `azure-sdk-for-net` repo, add the step to bump cadl emitter version

resolve #2902

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by pshao25** [comment]
So, could we remove the cadl-ci.yml?

**Comment by archerzz** [comment]
> So, could we remove the cadl-ci.yml?

Wait until we agree and verify the pipeline can work :)

**Comment by archerzz** [comment]
Here is the experimental build run:

- I manually triggered the build from PR branch: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2031414&view=results
- Here is the generated PR on `azure-sdk-for-net`: https://github.com/Azure/azure-sdk-for-net/pull/32866
  - Just in case the PR is force-pushed by subsequent change, here is the commit: [Azure/azure-sdk-for-net@`a62222d` (#32866)](https://github.com/Azure/azure-sdk-for-net/pull/32866/commits/a62222dc063a5f2d4df8844fe93f3693f670a474)
- As you can see, both [Packages.Data.props](https://github.com/Azure/azure-sdk-for-net/pull/32866/commits/a62222dc063a5f2d4df8844fe93f3693f670a474#diff-93a28d9569550c68624a8ad2209a6fae1d4f88237e4b5414eed2ecac4ef8c98f) and [csharp-emitter-package.json](https://github.com/Azure/azure-sdk-for-net/pull/32866/commits/a62222dc063a5f2d4df8844fe93f3693f670a474#diff-5e7d37047100b2bc3a83b37f06d0bb6076340a33e6479168b309de300a5e5395) are updated.
- Corresponding npm package [@autorest/csharp@3.0.0-beta.20221204.2](https://www.npmjs.com/package/@autorest/csharp/v/3.0.0-beta.20221204.2)
- Corresponding npm package [@azure-tools/cadl-csharp@3.0.0-beta.20221204.2](https://www.npmjs.com/package/@azure-tools/cadl-csharp/v/0.1.10-beta.20221204.2)
  - which also depends on `@autorest/csharp@3.0.0-beta.20221204.2` in its **package.json**

> "@autorest/csharp": "3.0.0-beta.20221204.2"

File: eng/pipelines/build.yml [hunk 1]
```diff
@@ -145,6 +156,7 @@ stages:
             vmImage: ubuntu-20.04
           variables:
             AutorestCSharpVersion: $[stageDependencies.Build_and_Test.Build.outputs['Package.AutorestCSharpVersion']]
+            CadlEmitterVersion: $[stageDependencies.Build_and_Test.Build.outputs['Publish.CadlEmitterVersion']]
```
Comment (chunyu3) [inline eng/pipelines/build.yml:159]: Where is to define CadlEmitterVersion? it is better that the minVersion of caldEmitterVersion is the same as autorest.csharp. such as 20221202.1

Comment (archerzz) [reply eng/pipelines/build.yml:159]: @chunyu3 It's defined in [PublishRelease.ps1](https://github.com/Azure/autorest.csharp/pull/2914/files#diff-3a01bf31f85915a7085de14ff3ccce2836de59f0acf8a98696edfa9d545580d2R53):

> Write-Host "##vso[task.setvariable variable=CadlEmitterVersion;isoutput=true]$devVersion"

Its suffix is the same as `autorest.csharp`, e.g. `beta.20211202.1`, see the updated [csharp-emitter-package.json](https://github.com/Azure/azure-sdk-for-net/pull/32866/files#diff-5e7d37047100b2bc3a83b37f06d0bb6076340a33e6479168b309de300a5e5395):

> "@azure-tools/cadl-csharp": "0.1.10-beta.20221204.2"

**Review by chunyu3 — APPROVED** [review]

**Pull request merged**